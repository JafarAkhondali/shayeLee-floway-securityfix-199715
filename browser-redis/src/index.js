import { variables } from "shaye-sword";

let cacheData = {};


class Redis {
  /**
   * 初始化redis
   * @param {object} initialData - 初始缓存数据
  */
  constructor(initialData) {
    if (variables.isCorrect(initialData)) cacheData = initialData;
  }

  /**
   * 存储分页数据
   * @param {object} params
   * @param {string} params.key - 缓存名称
   * @param {number} params.pageNum - 分页页码
   * @param {number} params.pageSize - 单页条数
   * @param {array} data
  */
  storePagingData = (params, data) => {
    if (!variables.isObject(params)) {
      throw new TypeError(`params expect a object, but got ${typeof params}.`);
    } else {
      const keyType = typeof params.key;
      const pageNumType = typeof params.pageNum;
      const pageSizeType = typeof params.pageSize;
      if (keyType !== "string") {
        throw new TypeError(`params.key expect a string, but got ${keyType}.`);
      } else if (pageNumType !== "number") {
        throw new TypeError(`params.pageNum expect a number, but got ${pageNumType}.`);
      } else if (pageSizeType !== "number") {
        throw new TypeError(`params.pageSize expect a number, but got ${pageSizeType}.`);
      }
    }

    if (!Array.isArray(data)) {
      throw new TypeError(`data expect a array, but got ${typeof data}.`);
    }    

    const { key, pageNum, pageSize } = params;
    const startCursor = (pageNum - 1) * pageSize;
    const endCursor = startCursor + (data.length - 1);
    if (!variables.isObject(cacheData[key])) {
      cacheData[key] = {
        startCursor: null,
        endCursor: null,
        data: []
      };
    }
    
    if (cacheData[key]["data"]["length"] === 0) {
      cacheData[key]["data"] = data;
      cacheData[key]["startCursor"] = startCursor;
      cacheData[key]["endCursor"] = endCursor;
    } else {
      const sourceArr = cacheData[key]["data"];
      if (
        (startCursor < cacheData[key]["startCursor"]) &&
        (endCursor >= cacheData[key]["startCursor"]) &&
        (endCursor <= cacheData[key]["endCursor"])
      ) {
        const newArr = data.slice(0, cacheData[key]["startCursor"] - startCursor);
        for (let i = (newArr.length - 1); i > -1; i--) {
          sourceArr.unshift(newArr[i]);
        }
        cacheData[key]["startCursor"] = startCursor;
      } else if (
        (startCursor < cacheData[key]["startCursor"]) &&
        (endCursor === (cacheData[key]["startCursor"] - 1))
      ) {
        const newArr = data;
        for (let i = (newArr.length - 1); i > -1; i--) {
          sourceArr.unshift(newArr[i]);
        }
        cacheData[key]["startCursor"] = startCursor;
      } else if (
        (startCursor < cacheData[key]["startCursor"]) &&
        (endCursor > cacheData[key]["endCursor"])
      ) {
        const leftArr = data.slice(0, cacheData[key]["startCursor"] - startCursor);
        for (let i = (leftArr.length - 1); i > -1; i--) {
          sourceArr.unshift(leftArr[i]);
        }
        cacheData[key]["startCursor"] = startCursor;
        const rightArr = data.slice(cacheData[key]["endCursor"] - endCursor);
        rightArr.forEach(item => {
          sourceArr.push(item);
        })
        cacheData[key]["endCursor"] = endCursor;
      } else if (
        (startCursor >= cacheData[key]["startCursor"]) &&
        (startCursor <= cacheData[key]["endCursor"]) &&
        (endCursor > cacheData[key]["endCursor"])
      ) {
        const newArr = data.slice(cacheData[key]["endCursor"] - endCursor);
        newArr.forEach(item => {
          sourceArr.push(item);
        });
        cacheData[key]["endCursor"] = endCursor;
      } else if (startCursor === (cacheData[key]["endCursor"] + 1)) {
        const newArr = data;
        for (let i = (newArr.length - 1); i > -1; i--) {
          sourceArr.push(newArr[i]);
        }
        cacheData[key]["endCursor"] = endCursor;
      }
    }
  }
}

export default Redis;