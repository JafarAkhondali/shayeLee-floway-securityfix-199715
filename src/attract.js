import { isCorrectVal, isObject } from "./utils";
import store from "./store";
const eventLog = store.eventLog;
import { Subject, of } from "rxjs";
import { pluck, filter, switchMap } from "rxjs/operators";
import { pipeFromArray } from "rxjs/internal/util/pipe";
import md5 from "js-md5";
import eventBus$ from "./eventBus";
import schema from "async-validator";

/**
 * 捕捉事件触发器(distributor$)推送的指定类型的事件，并返回一个observable
 * @param {string|object} type - 事件类型 或 事件对象
 */
const attract = function(type) {
  const validator = new schema({
    type(rule, value, callback) {
      const errors = [];
      if (!(typeof type === "string" || isObject(type))) {
        errors.push(new Error("type must be string or object"));
      }
      callback(errors);
    }
  });
  validator.validate({ type }, errors => {
    if (errors) {
      console.error(errors[0]);
    }
  });

  let options = null;
  const _options = {
    useCache: false,
    cacheType: "eventCache" // eventCache itemCache
  };

  if (isCorrectVal(type) && typeof type === "string") {
    options = Object.assign({}, _options);
  } else if (isCorrectVal(type) && isObject(type)) {
    options = Object.assign({}, _options, type);
    type = options.type;
    delete options.type;
  }

  const event$ = eventBus$.pipe(
    pluck(type),
    filter(event => {
      if (!isCorrectVal(event)) return false;

      if (!isCorrectVal(eventLog.pushHeadersMap[event.type])) {
        eventLog.pushHeadersMap[event.type] = {
          event,
          lastModifyId: new Date().getTime()
        };
        return true;
      }

      const pushHeaders = eventLog.pushHeadersMap[event.type];
      const lastEvent = pushHeaders.event;

      // 判断是否要更新lastModifyId
      if (
        !options.useCache ||
        (
          md5(JSON.stringify(lastEvent.payload)) !== md5(JSON.stringify(event.payload)) ||
          md5(JSON.stringify(lastEvent.options)) !== md5(JSON.stringify(event.options))
        )
      ) {
        eventLog.pushHeadersMap[event.type][
          "lastModifyId"
        ] = new Date().getTime();
      }
      pushHeaders.event = event;
      return true;
    })
  );

  const operations = [];
  let _subscription = {
    unsubscribe: function() {}
  };
  function generateObs(obs$) {
    _subscription.unsubscribe();
    const obs$$ = new Subject();
    obs$$.__type__ = type;
    const _obs$ = obs$.pipe(
      switchMap(event => {
        const pushHeaders = eventLog.pushHeadersMap[event.type];
        let hasModified = obs$$.lastModifyId !== pushHeaders.lastModifyId;

        // 判断是否有缓存数据
        let cacheData;
        if (options.useCache && !hasModified) {
          switch (options.cacheType) {
            case "eventCache":
              cacheData = eventLog.dataMap[event.type];
              if (!isCorrectVal(cacheData)) {
                hasModified = true;
                pushHeaders.lastModifyId = new Date().getTime();
              }
              break;
          }
        }
        event.hasModified = hasModified;
        return hasModified
          ? operations.length === 0
            ? of(event)
            : pipeFromArray(operations)(of(event))
          : of(cacheData);
      }),
      filter(data => {
        const canPass = !(data === null || typeof data === "undefined");
        const pushHeaders = eventLog.pushHeadersMap[type];
        const event = pushHeaders.event;
        const hasModified = event.hasModified;
        if (canPass) {
          obs$$.lastModifyId = pushHeaders.lastModifyId;
        }

        // 缓存数据
        if (canPass && hasModified) {
          switch (options.cacheType) {
            case "eventCache":
              eventLog.dataMap[type] = data;
              break;
          }
        }

        return canPass;
      })
    );
    _subscription = _obs$.subscribe(obs$$);
    return obs$$;
  }
  const processEvent$ = generateObs(event$);

  processEvent$.pipe = function() {
    for (var i = 0; i < arguments.length; i++) {
      operations.push(arguments[i]);
    }
    return generateObs(event$);
  };
  return processEvent$;
};
export default attract;
