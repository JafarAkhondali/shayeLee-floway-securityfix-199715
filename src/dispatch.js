import store from './store';
const producerMap = store.producerMap;

const dispatch = function(stateName, action) {
  /* if (!Array.isArray(actions)) {
    actions = [actions];
  }
  const map = {};
  actions.forEach(action => {
    if (typeof action === "string") {
      const type = action;
      action = { type };
    }
    action.type = `${stateName}#${action.type}`;

    map[action.type] = action;
  });
  eventBus.next(map); */

  if (typeof action === "string") {
    const type = action;
    action = { type };
  }

  producerMap[stateName](action);
}

export default dispatch;
