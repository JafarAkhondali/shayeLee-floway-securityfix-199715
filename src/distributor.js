/**
 * 全局事件触发器
 * @namespace distributor$
*/

import eventBus$ from "./eventBus";
import { isCorrectVal } from "./utils";

const distributor$ = {
  /**
   * 推送事件
   * @memberof distributor$
   * @method next
   * @param {string|object|objectArray} events - 事件集合
   * @example
   * import { distributor$ } from "rx-samsara";
   * distributor$.next("eventName");
   * @example
   * import { distributor$ } from "rx-samsara";
   * distributor$.next({
   *   type: "eventName"
   * });
   * @example
   * import { distributor$ } from "rx-samsara";
   * distributor$.next([
   *   {
   *     type: "eventName"
   *   }
   * ]);
  */
  next: function(events) {
    if (!Array.isArray(events)) {
      events = [events];
    }
    const map = {};
    events.forEach(event => {
      if (typeof event === "string") {
        const type = event;
        event = { type: type };
      }
      if (!isCorrectVal(event.payload)) event.payload = {};
      if (!isCorrectVal(event.options)) event.options = {};
      map[event.type] = event;
    });
    eventBus$.next(map);
  }
};
export default distributor$;
