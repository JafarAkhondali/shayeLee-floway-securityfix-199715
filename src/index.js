import config from "./config";

const _setConfig = function(customConfig) {
  Object.keys(customConfig).forEach(key => {
    config[key] = customConfig[key];
  });
};
export const setConfig = _setConfig;

export { default as State } from "./state";

export { default as ofHot } from "./ofHot";

export { default as distributor$ } from "./distributor";

export { default as ofLast } from "./ofLast";

export { default as attract } from "./attract";

export { default as permeate } from "./permeate";
