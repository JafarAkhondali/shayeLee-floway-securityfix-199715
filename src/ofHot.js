import { BehaviorSubject } from "rxjs";
import { isObject } from "./utils";

/**
 * 创建一个包含指定数据、并且可更新数据的observable
 * @function hotObservable
 * @param {*} data
 */
function ofHot(data) {
  const state$ = new BehaviorSubject(data);
  state$.__data__ = data;
  state$.update = function(mData) {
    let newData;
    const _data = typeof mData === "function" ? mData(state$.__data__) : mData;
    if (isObject(_data)) {
      newData = Object.assign({}, state$.__data__, _data);
    } else {
      newData = _data;
    }
    state$.__data__ = newData;
    state$.next(newData);
  };
  return state$;
}
export default ofHot;
