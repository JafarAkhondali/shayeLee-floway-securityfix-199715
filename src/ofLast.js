import { map, combineLatest } from "rxjs/operators";

const ofLast = function(obs$) {
  return obs$.pipe(
    combineLatest(),
    map(arr => arr[0])
  );
};
export default ofLast;
