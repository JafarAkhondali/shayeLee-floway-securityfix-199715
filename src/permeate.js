import React from "react";
import ofHot from "./ofHot";
import rxStore from "./store";
import { isObject, isCorrectVal } from "./utils";

/**
 * observable与react组件的集成(将observable转换为组件属性)
 * @param {object} observablesMap - 可观察对象集合
 * @param {object} inputOptions - 选项
 * @param {object} inputOptions.defaultProps - 组件的默认属性
 * @param {array} inputOptions.delayeringFields - 推送数据需要扁平化的可观察对象的key
 * @example
 * import { permeate } from "rx-samsara";
 * import { from } from "rxjs";
 * 
 * const testData$ = from(
 *   new Promise(resolve => {
 *     setTimeout(() => {
 *       resolve(123);
 *     }, 800);
 *   })
 * };
 * 
 * class CompA extends React.Component {
 *   render() {
 *     // this.props.testData === 123
 *     return <div>{this.props.testData}</div>;
 *   }
 * }
 * 
 * export default permeate({ testData: testData$ })(CompA);
*/
const permeate = function(observablesMap, inputOptions) {
  let options = inputOptions || {};

  const handler = function(Comp) {
    if (!isObject(observablesMap))
      throw new TypeError(`方法permeate()的参数observablesMap必须是object类型`);

    class Permeate extends React.PureComponent {
      constructor() {
        super();

        this.state = {};
        this.subscriptionArr = [];
        this.suspendedObservableKeys = Object.keys(observablesMap);
        this._suspendedObservables = [];
        this._innerObservableMaps = {};
        if (this.suspendedObservableKeys.length > 0) {
          this.suspendedObservableKeys.forEach(key => {
            let _observable;
            if (typeof observablesMap[key]["subscribe"] !== "function") {
              _observable = ofHot(observablesMap[key]);
              this._innerObservableMaps[`${key}$`] = _observable;
              this.state[key] = observablesMap[key];
            } else {
              _observable = observablesMap[key];
            }
            this._suspendedObservables.push(_observable);
          });
        } else {
          throw new TypeError(
            `方法permeate()的参数observablesMap不允许传一个空的object`
          );
        }

        this.state = Object.assign(
          {},
          this._innerObservableMaps,
          isCorrectVal(options.defaultProps) ? options.defaultProps : {}
        );
      }

      componentWillMount() {
        const obsArr = this._suspendedObservables,
          len = obsArr.length;
        for (let i = 0; i < len; i++) {
          const subscription = obsArr[i].subscribe(data => {
            const type = obsArr[i]["__type__"];
            const pushHeaders = rxStore.pushHeadersMap[type];

            if (
              options.delayeringFields &&
              options.delayeringFields.includes(this.suspendedObservableKeys[i])
            ) {
              const _stateObj = {};
              for (const key in data) {
                if (this.state[key] !== data[key]) {
                  _stateObj[key] = data[key];
                }
              }
              // if (isCorrectVal(pushHeaders))  console.log(pushHeaders);
              this.setState(_stateObj);
              return;
            }

            if (this.state[this.suspendedObservableKeys[i]] !== data) {
              // if (isCorrectVal(pushHeaders))  console.log(pushHeaders);
              this.setState({
                [this.suspendedObservableKeys[i]]: data
              });
            }
          });
          this.subscriptionArr.push(subscription);
        }
      }
      componentWillUnmount() {
        this.subscriptionArr.forEach(subscription => {
          subscription.unsubscribe();
        });
      }
      render() {
        return <Comp {...this.props} {...this.state} />;
      }
    }
    Permeate.displayName = `Permeate(${Comp.displayName ||
      Comp.name ||
      "Component"})`;
    return Permeate;
  };
  return handler;
};
export default permeate;
