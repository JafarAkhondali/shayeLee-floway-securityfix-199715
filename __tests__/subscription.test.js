import React from 'react';
import renderer from 'react-test-renderer';
import { subscription, state } from "../src/index";

const str$ = state({
  name: "str",
  defaultValue: "floway"
});

@subscription({
  str: str$
})
class TestComp extends React.Component {
  render() {
    return (
      <div>{this.props.str}</div>
    )
  }
}

test("test @subscription", () => {
  const component = renderer.create(
    <TestComp />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("test @subscription with empty object param", () => {
  expect(() => {
    @subscription({})
    class Comp extends React.Component {
      render() {
        return (
          <div></div>
        )
      }
    }

    renderer.create(
      <Comp />
    );
  }).toThrow();
});

test("test @subscription with erring param", () => {
  expect(() => {
    subscription([str$])(function (props) {
      return (
        <div>props.str</div>
      )
    })
  }).toThrow();
});