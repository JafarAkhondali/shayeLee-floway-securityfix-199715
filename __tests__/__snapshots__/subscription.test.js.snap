// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test @subscription 1`] = `
<div>
  floway
</div>
`;
